main/maven/test/src/mill/main/maven/BuildGenTests.scala

```diff
@@ -125,15 +129,61 @@ object BuildGenTests extends TestSuite {
       .map(PathRef(_))
       .toSeq
 
-  def checkFiles(actualFiles: Seq[PathRef], expectedRoot: os.Path): Boolean = {
-    val expectedFiles = buildFiles(expectedRoot)
+  def checkFiles(actualFiles: Seq[os.SubPath], root: os.Path, expectedRoot: os.Path): Boolean = {
+    val expectedFiles = buildFiles(expectedRoot).map(_.path.relativeTo(expectedRoot).asSubPath)
+
+    val actualFilesSet = actualFiles.toSet
+    val expectedFilesSet = expectedFiles.toSet
+
+    val missing = expectedFiles.filterNot(actualFilesSet)
+    val extra = actualFiles.filterNot(expectedFilesSet)
 
-    actualFiles.nonEmpty &&
-    actualFiles.length == expectedFiles.length &&
-    actualFiles.iterator.zip(expectedFiles.iterator).forall {
-      case (actual, expected) =>
-        actual.path.endsWith(expected.path.relativeTo(expectedRoot)) &&
-        os.read.lines(actual.path) == os.read.lines(expected.path)
+    val shared = actualFiles.filter(expectedFilesSet)
+
+    val differentContent = shared.filter { subPath =>
+      val actual = os.read.lines(root / subPath)
+      val expected = os.read.lines(expectedRoot / subPath)
+      actual != expected
     }
+
+    val valid = missing.isEmpty && extra.isEmpty && differentContent.isEmpty
+
+    if (!valid)
+      if (updateSnapshots) {
+        System.err.println(
+          s"Expected and actual files differ, updating expected files in resources under $expectedRoot"
+        )
+
+        for (subPath <- missing) {
+          val path = expectedRoot / subPath
+          System.err.println(s"Removing $subPath")
+          os.remove(path)
+        }
+
+        for (subPath <- extra) {
+          val source = root / subPath
+          val dest = expectedRoot / subPath
+          System.err.println(s"Creating $subPath")
+          os.copy(source, dest, createFolders = true)
+        }
+
+        for (subPath <- differentContent) {
+          val source = root / subPath
+          val dest = expectedRoot / subPath
+          System.err.println(s"Updating $subPath")
+          os.copy.over(source, dest, createFolders = true)
+        }
```

Comment:
Could `valid` be computed using `git diff` as well? If `exitCode != 0`, then there is a difference between the expected and generated files, and the test should fail.

`updateSnapshots` seems like it could be implemented by a single coarse-grained `os.copy` of the entire generated folder to replace the expected folder.

Reply:
Just tried that, it seems to work.

I was using manual comparison and piecewise updates for two reasons: run sub-processes can be flaky sometimes, especially on Windows, and removing and copying all files updates last-modified times, which can sometimes lead to unintended side effects (with file watchers for example). But that might overcautious…

Reply:
Seems it's a problem on the CI, where we get [things](https://github.com/com-lihaoyi/mill/actions/runs/12255214110/job/34188017870?pr=4079#step:11:3344) like
```text
[11002-1] diff --git a/home/runner/work/mill/mill/main/maven/test/resources/expected/maven-samples/build.mill b/home/runner/work/mill/mill/out/main/maven/test/testCached.dest/mill.main.maven.BuildGenTests/sandbox/maven-samples/build.mill
[11002-1] old mode 100755
[11002-1] new mode 100644
[11002-1] diff --git a/home/runner/work/mill/mill/main/maven/test/resources/expected/maven-samples/multi-module/package.mill b/home/runner/work/mill/mill/out/main/maven/test/testCached.dest/mill.main.maven.BuildGenTests/sandbox/maven-samples/multi-module/package.mill
[11002-1] old mode 100755
[11002-1] new mode 100644
[11002-1] diff --git a/home/runner/work/mill/mill/main/maven/test/resources/expected/maven-samples/multi-module/server/package.mill b/home/runner/work/mill/mill/out/main/maven/test/testCached.dest/mill.main.maven.BuildGenTests/sandbox/maven-samples/multi-module/server/package.mill
[11002-1] old mode 100755
[11002-1] new mode 100644
[11002-1] diff --git a/home/runner/work/mill/mill/main/maven/test/resources/expected/maven-samples/multi-module/webapp/package.mill b/home/runner/work/mill/mill/out/main/maven/test/testCached.dest/mill.main.maven.BuildGenTests/sandbox/maven-samples/multi-module/webapp/package.mill
[11002-1] old mode 100755
[11002-1] new mode 100644
[11002-1] diff --git a/home/runner/work/mill/mill/main/maven/test/resources/expected/maven-samples/single-module/package.mill b/home/runner/work/mill/mill/out/main/maven/test/testCached.dest/mill.main.maven.BuildGenTests/sandbox/maven-samples/single-module/package.mill
[11002-1] old mode 100755
[11002-1] new mode 100644
```

Reply:
How about using os.walk/os.perms.set to normalize the permissions before running git diff? should be a one liner more or less